contracts/PillarStaking.sol

```diff
@@ -0,0 +1,281 @@
+        if ((totalStaked + _amount) > maxTotalStake)
+            revert MaximumTotalStakeReached({
+                totalMaxStake: maxTotalStake,
+                currentStakedAmount: totalStaked,
+                remainingStakeableAmount: (maxTotalStake - totalStaked),
+                stakerAmount: _amount
+            });
+        if (_amount < minStake)
+            revert InvalidMinimumStake({minimumStakeAmount: minStake});
+        if (_amount > maxStake)
+            revert InvalidMaximumStake({maximumStakeAmount: maxStake});
+        if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
+            revert StakeWouldBeGreaterThanMax();
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
+            stakeholderData[msg.sender].stakedAmount +
+            _amount;
+        if (!stakeholderData[msg.sender].staked) {
+            stakeholderList.push(msg.sender);
+        }
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        emit Staked(msg.sender, _amount);
+    }
+
+    function unstake()
+        external
+        nonReentrant
+        whenReadyForUnstake
+        whenRewardsAllocated
+    {
+        uint256 stakedBalance = stakeholderData[msg.sender].stakedAmount;
+        totalStaked = totalStaked - stakedBalance;
+        stakeholderData[msg.sender].stakedAmount = 0;
+        stakingToken.safeTransfer(msg.sender, stakedBalance);
+        uint256 rewardAmount = stakeholderData[msg.sender].rewardAmount;
+        stakeholderData[msg.sender].rewardAmount = 0;
+        rewardToken.safeTransfer(msg.sender, rewardAmount);
+        emit Unstaked(msg.sender, stakedBalance);
+        emit RewardPaid(msg.sender, rewardAmount);
+    }
+
+    /* ========== VIEWS ========== */
+
+    function getContractState() public view returns (StakingState) {
+        return state;
+    }
+
+    function getStakedAmountForAccount(address account)
+        public
+        view
+        returns (uint256 amount)
+    {
+        return stakeholderData[account].stakedAmount;
+    }
+
+    function getRewardAmountForAccount(address account)
+        external
+        view
+        returns (uint256 amount)
+    {
+        return stakeholderData[account].rewardAmount;
+    }
+
+    /* ========== RESTRICTED FUNCTIONS ========== */
+
+    function depositRewards(uint256 _amount) external onlyOwner {
+        if (_amount == 0) revert RewardsCannotBeZero();
+        rewardToken.safeTransferFrom(msg.sender, address(this), _amount);
+        rewardsDeposited = true;
+        emit RewardsDeposited(_amount);
+    }
+
+    function calculateRewardAllocation() external onlyOwner {
+        if (!rewardsDeposited) revert RewardsNotTransferred();
+        uint256 totalRewards = rewardToken.balanceOf(address(this));
+        for (uint256 i; i < stakeholderList.length; ++i) {
```

Comment:
Or maybe we should add the `claim` function, so users will pay for the calculation of their rewards themselves?

Reply:
I suppose I could separate this to work for a specific address and have that built into the unstake fn. I will need to rethink how the logic will work for this

Reply:
What I am thinking is rather than calculating rewards for all users in one function, have a function that calculates the rewards for 1 user and build this into the unstake function so that the user will pay for the calculation of their own rewards. This amount will be stored against the user so that they can see how much they should be paid. The only issue with this is that before unstaking (and paying out of rewards), the user will not know how much rewards are due to them

Reply:
>have a function that calculates the rewards for 1 user and build this into the unstake function so that the user will pay for the calculation of their own rewards

Yeah, I think this would be better.

>The only issue with this is that before unstaking (and paying out of rewards), the user will not know how much rewards are due to them

I'm not quite getting this part 🤔 Rewards are fixed after the lock period begins, no?

Reply:
No as we dont know what the amount they are due would be as the rewards are unknown. Their % of rewards would be know but not the amount.

Reply:
This is a good point @marie-fourier . I would say the reward calculation logic should be part of the unstake function. We determine how much WETH a user is eligible for at the time of unstake and let that be transferred to him.